File: ponte/src/catalogo.js
/* Il catalogo di integrazioni, dispositivi ed entita' di Home Assistant.
 *
 * Un elettrodomestico moderno arriva in Home Assistant da un'integrazione —
 * hOn per Hoover, Candy e Haier, Home Connect per Bosch e Siemens, Miele, LG
 * ThinQ — come un *dispositivo* con dentro venti o trenta entita': il
 * programma, la fase, il tempo che manca, l'energia del ciclo, l'oblo'. Lo
 * stesso vale per un'auto o per un robot. La plancia li sceglie da un menu
 * — l'integrazione, il dispositivo, e dentro le sue entita' — e quel menu lo
 * costruiva l'integrazione di Home Assistant leggendo i registri. Qui lo
 * costruisce il ponte, allo stesso modo e con la stessa forma: la pagina non
 * deve accorgersi della differenza.
 *
 * Il ponte non ha i registri in mano come li aveva l'integrazione: li chiede
 * a Home Assistant sul suo filo — `config/device_registry/list`,
 * `config/entity_registry/list`, `config/area_registry/list`,
 * `config_entries/get`, `manifest/list` — e li tiene per mezzo minuto, perche'
 * il menu li chiede piu' volte di seguito e i registri non cambiano a ogni
 * tocco. Legge e basta: non scrive niente.
 *
 * Le entita' si chiedono per dispositivo e non tutte insieme: una casa grande
 * ne ha migliaia, e il menu ne vuole vedere venti alla volta.
 */

/* Piu' di cosi' non e' un menu: e' un'esportazione. */
export const DISPOSITIVI_MASSIMI = 200;

/* Le entita' della plancia stessa non sono di nessun elettrodomestico. */
const DOMINIO_DELLA_PLANCIA = "dashboardmodern";

const testo = (valore) => (valore == null ? "" : String(valore).trim());

/* Il nome dell'entita' senza quello del dispositivo davanti.
 *
 * Un'integrazione moderna compone «Lavatrice Tempo rimanente» dal nome del
 * dispositivo e da quello suo: nel menu del dispositivo la prima meta' e'
 * gia' scritta in cima, e ripeterla venti volte non aiuta a leggere. */
function nomeDellEntita(voce, stato, nomeDelDispositivo) {
  const proprio = testo(voce.name) || testo(voce.original_name);
  if (proprio) return proprio;
  const amichevole = testo(stato?.attributes?.friendly_name);
  if (
    amichevole &&
    nomeDelDispositivo &&
    amichevole.toLowerCase().startsWith(nomeDelDispositivo.toLowerCase())
  ) {
    const resto = amichevole.slice(nomeDelDispositivo.length).replace(/^[\s\-:·]+|[\s\-:·]+$/g, "");
    if (resto) return resto;
  }
  if (amichevole) return amichevole;
  return String(voce.entity_id || "")
    .split(".")
    .slice(1)
    .join(".")
    .replace(/_/g, " ");
}

/* Una riga del catalogo per un'entita'. */
function rigaDellEntita(voce, stato, nomeDelDispositivo) {
  const attributi = stato?.attributes || {};
  return {
    entity_id: voce.entity_id,
    device_id: voce.device_id || "",
    platform: voce.platform,
    name: nomeDellEntita(voce, stato, nomeDelDispositivo),
    translation_key: testo(voce.translation_key),
    device_class:
      testo(voce.device_class) ||
      testo(voce.original_device_class) ||
      testo(attributi.device_class),
    unit: testo(voce.unit_of_measurement) || testo(attributi.unit_of_measurement),
    state_class: testo(attributi.state_class),
    category: testo(voce.entity_category),
    disabled: voce.disabled_by != null,
    hidden: voce.hidden_by != null,
  };
}

/* Una riga del catalogo per un'integrazione.
 *
 * `custom` e' vero per un'integrazione che sta in `custom_components` — da
 * HACS o copiata a mano — falso per una di Home Assistant, e `null` quando
 * il manifesto non si trova piu': le sue entita' restano nei registri anche
 * dopo che qualcuno l'ha tolta. */
function rigaDellIntegrazione(dominio, manifesto, voci) {
  return {
    domain: dominio,
    name: (manifesto && testo(manifesto.name)) || dominio,
    custom: manifesto ? manifesto.is_built_in !== true : null,
    entries: voci.map((voce) => ({
      entry_id: voce.entry_id,
      title: testo(voce.title),
      state: testo(voce.state),
    })),
    devices: 0,
  };
}

const perNome = (uno, altro) => {
  const a = uno.name.toLowerCase();
  const b = altro.name.toLowerCase();
  if (a < b) return -1;
  if (a > b) return 1;
  return uno.id < altro.id ? -1 : uno.id > altro.id ? 1 : 0;
};

const elenco = (valore) => (Array.isArray(valore) ? valore : []);

/* Integrazioni, dispositivi e — per i dispositivi chiesti — le entita'.
 *
 * Un dispositivo senza entita' non compare: non c'e' niente da mostrare. E'
 * una funzione pura dei registri, e si prova cosi'. */
export function costruisciIlCatalogo({
  dispositivi,
  entita,
  aree,
  voci,
  stati,
  manifesti,
  deviceIds = null,
} = {}) {
  const perDispositivo = new Map();
  for (const voce of elenco(entita)) {
    if (!voce || voce.platform === DOMINIO_DELLA_PLANCIA || !voce.device_id) continue;
    if (!perDispositivo.has(voce.device_id)) perDispositivo.set(voce.device_id, []);
    perDispositivo.get(voce.device_id).push(voce);
  }

  const dominioDellaVoce = new Map();
  for (const voce of elenco(voci)) {
    if (voce?.entry_id && voce.domain) dominioDellaVoce.set(voce.entry_id, voce.domain);
  }
  const nomeDellArea = new Map();
  for (const area of elenco(aree)) {
    if (area?.area_id) nomeDellArea.set(area.area_id, testo(area.name));
  }

  const devices = [];
  const domini = new Set();
  for (const dispositivo of elenco(dispositivi)) {
    const sue = perDispositivo.get(dispositivo?.id);
    if (!sue) continue;
    const piattaforme = new Map();
    for (const voce of sue)
      piattaforme.set(voce.platform, (piattaforme.get(voce.platform) || 0) + 1);
    let principale = "";
    if (dispositivo.primary_config_entry)
      principale = dominioDellaVoce.get(dispositivo.primary_config_entry) || "";
    if (!principale) {
      principale = [...piattaforme.entries()].sort((a, b) => b[1] - a[1])[0][0];
    }
    const integrazioni = [
      ...new Set([
        ...piattaforme.keys(),
        ...elenco(dispositivo.config_entries)
          .filter((entryId) => dominioDellaVoce.has(entryId))
          .map((entryId) => dominioDellaVoce.get(entryId)),
      ]),
    ]
      .filter((dominio) => dominio && dominio !== DOMINIO_DELLA_PLANCIA)
      .sort();
    if (!integrazioni.length) continue;
    if (!integrazioni.includes(principale)) principale = integrazioni[0];
    for (const dominio of integrazioni) domini.add(dominio);
    devices.push({
      id: dispositivo.id,
      name: testo(dispositivo.name_by_user) || testo(dispositivo.name),
      manufacturer: testo(dispositivo.manufacturer),
      model: testo(dispositivo.model) || testo(dispositivo.model_id),
      integration: principale,
      integrations: integrazioni,
      area_id: dispositivo.area_id || "",
      area: dispositivo.area_id ? nomeDellArea.get(dispositivo.area_id) || "" : "",
      entities: sue.filter((voce) => voce.disabled_by == null).length,
      disabled: dispositivo.disabled_by != null,
    });
  }
  devices.sort(perNome);

  const perDominio = new Map();
  for (const manifesto of elenco(manifesti)) {
    if (manifesto?.domain) perDominio.set(manifesto.domain, manifesto);
  }
  const integrations = [...domini].sort().map((dominio) =>
    rigaDellIntegrazione(
      dominio,
      perDominio.get(dominio) || null,
      elenco(voci).filter((voce) => voce?.domain === dominio),
    ),
  );
  const conteggio = new Map();
  for (const device of devices) {
    conteggio.set(device.integration, (conteggio.get(device.integration) || 0) + 1);
  }
  for (const integrazione of integrations) {
    integrazione.devices = conteggio.get(integrazione.domain) || 0;
  }
  integrations.sort((a, b) => {
    const x = a.name.toLowerCase();
    const y = b.name.toLowerCase();
    return x < y ? -1 : x > y ? 1 : 0;
  });

  const entities = [];
  if (Array.isArray(deviceIds) && deviceIds.length) {
    const statoDi = new Map();
    for (const stato of elenco(stati)) if (stato?.entity_id) statoDi.set(stato.entity_id, stato);
    const nomi = new Map(devices.map((device) => [device.id, device.name]));
    for (const deviceId of deviceIds.slice(0, DISPOSITIVI_MASSIMI)) {
      const sue = [...(perDispositivo.get(deviceId) || [])].sort((a, b) =>
        a.entity_id < b.entity_id ? -1 : a.entity_id > b.entity_id ? 1 : 0,
      );
      for (const voce of sue) {
        entities.push(rigaDellEntita(voce, statoDi.get(voce.entity_id), nomi.get(deviceId) || ""));
      }
    }
  }

  return { integrations, devices, entities };
}

/* Quanto si tengono i registri prima di richiederli. */
const QUANTO_DURANO = 30_000;

export class Catalogo {
  constructor({ casa, registro, adesso = () => Date.now(), quantoDurano = QUANTO_DURANO } = {}) {
    this.casa = casa;
    this.registro = registro ?? { info() {}, attenzione() {}, errore() {} };
    this.adesso = adesso;
    this.quantoDurano = quantoDurano;
    this._registri = null;
    this._registriLettiIl = 0;
    this._stati = null;
    this._statiLettiIl = 0;
  }

  /* Il catalogo, nella forma che la pagina si aspetta. Solleva se Home
   * Assistant non risponde alle domande che non possono mancare. */
  async chiedi({ deviceIds = null } = {}) {
    const registri = await this._iRegistri();
    const stati = Array.isArray(deviceIds) && deviceIds.length ? await this._gliStati() : [];
    return costruisciIlCatalogo({ ...registri, stati, deviceIds });
  }

  async _iRegistri() {
    const ora = this.adesso();
    if (this._registri && ora - this._registriLettiIl < this.quantoDurano) return this._registri;
    const [dispositivi, entita, aree, voci, manifesti] = await Promise.all([
      this.casa.chiedi({ type: "config/device_registry/list" }),
      this.casa.chiedi({ type: "config/entity_registry/list" }),
      this._facoltativa({ type: "config/area_registry/list" }),
      /* Le voci di configurazione e i manifesti vogliono un amministratore, e
       * il segno del Supervisor lo e'; se un giorno non lo fosse, il catalogo
       * esce lo stesso, con i nomi delle integrazioni al posto dei nomi belli. */
      this._facoltativa({ type: "config_entries/get" }),
      this._facoltativa({ type: "manifest/list" }),
    ]);
    this._registri = {
      dispositivi: elenco(dispositivi),
      entita: elenco(entita),
      aree: elenco(aree),
      voci: elenco(voci),
      manifesti: elenco(manifesti),
    };
    this._registriLettiIl = this.adesso();
    return this._registri;
  }

  async _gliStati() {
    const ora = this.adesso();
    if (this._stati && ora - this._statiLettiIl < this.quantoDurano) return this._stati;
    this._stati = elenco(await this._facoltativa({ type: "get_states" }));
    this._statiLettiIl = this.adesso();
    return this._stati;
  }

  async _facoltativa(comando) {
    try {
      return await this.casa.chiedi(comando);
    } catch (errore) {
      this.registro.attenzione(`${comando.type} non ha risposto: ${errore?.message || errore}`);
      return [];
    }
  }
}

File: app/lib/casa/catalogo/catalogo.dart
/// Il catalogo delle integrazioni: cosa c'e' collegato a Home Assistant.
///
/// Un elettrodomestico moderno non e' un'entita': e' **un dispositivo con
/// dentro venti o trenta entita'**. hOn per Hoover, Candy e Haier; Home
/// Connect per Bosch e Siemens; Miele; LG ThinQ. Lo stesso vale per un'auto o
/// per un robot: il programma, la fase, il tempo che manca, l'energia del
/// ciclo, l'oblo', ognuno per conto suo.
///
/// Chi configura la sua lavatrice non deve battere venti identificativi: deve
/// scegliere l'integrazione, poi il dispositivo, e le entita' le trova gia'
/// li'. E' quello che fa la Config della dashboard, ed e' il motivo per cui
/// configurare l'auto elettrica a mano — diciassette caselle — non e' una
/// cosa che si chiede a qualcuno.
///
/// Il pezzo difficile e' gia' fatto e sta nel ponte (`ponte/src/catalogo.js`):
/// legge i registri di Home Assistant e serve il menu con la stessa forma che
/// aveva l'integrazione. Qui c'e' solo chi lo chiede.
library;

import '../../ponte/filo.dart';

/// Quanto si aspetta il catalogo: legge quattro registri e li mette insieme.
const attesaDelCatalogo = Duration(seconds: 25);

/// Il comando del ponte.
const comandoDelCatalogo = 'dashboardmodern/integrations/catalog';

/// Un'integrazione: hOn, Home Connect, Miele.
class Integrazione {
  const Integrazione({
    required this.dominio,
    required this.nome,
    required this.quantiDispositivi,
    required this.diQualcunAltro,
  });

  /// Come si chiama nel codice: `hon`, `home_connect`.
  final String dominio;

  /// Come si chiama per chi guarda: «hOn», «Home Connect».
  final String nome;

  final int quantiDispositivi;

  /// `true` quando non e' una delle integrazioni di serie di Home Assistant:
  /// e' arrivata da HACS o da una cartella. Non cambia niente, ma chi cerca la
  /// sua lavatrice e non la trova vuole sapere se l'integrazione e' quella
  /// giusta.
  final bool diQualcunAltro;
}

/// Un dispositivo: la lavatrice, l'auto, il robot.
class Dispositivo {
  const Dispositivo({
    required this.id,
    required this.nome,
    required this.marca,
    required this.modello,
    required this.integrazione,
    required this.stanza,
    required this.quanteEntita,
    required this.spento,
  });

  final String id;
  final String nome;
  final String marca;
  final String modello;
  final String integrazione;
  final String stanza;
  final int quanteEntita;

  /// Disabilitato in Home Assistant: si vede, ma le sue entita' non
  /// risponderanno.
  final bool spento;

  /// Marca e modello in una riga, per chi ha tre lavatrici uguali di nome.
  String get sotto => [
    if (marca.isNotEmpty) marca,
    if (modello.isNotEmpty) modello,
    if (stanza.isNotEmpty) stanza,
  ].join(' · ');
}

/// Un'entita' di un dispositivo.
class EntitaDelDispositivo {
  const EntitaDelDispositivo({
    required this.id,
    required this.nome,
    required this.classe,
    required this.unita,
    required this.categoria,
    required this.spenta,
    this.chiaveDiTraduzione = '',
    this.classeDiStato = '',
    this.piattaforma = '',
  });

  final String id;

  /// Il nome **senza quello del dispositivo davanti**: nel menu di una
  /// lavatrice «Lavatrice Tempo rimanente» diventa «Tempo rimanente», che e'
  /// l'unica meta' che serve a distinguerlo dalle altre diciannove.
  final String nome;

  final String classe;
  final String unita;

  /// Come l'integrazione chiama questa entita' **in ogni lingua**.
  ///
  /// E' l'indizio piu' fedele che ci sia: hOn chiama `remaining_time` il tempo
  /// che manca sia su un Home Assistant in italiano sia su uno in coreano,
  /// mentre il nome cambia. Il motore che indovina le caselle la pesa quanto
  /// il nome e l'identificativo messi insieme.
  final String chiaveDiTraduzione;

  /// Di che integrazione e': `proxmoxve`, `fritz`. Lo stato non lo dice — la
  /// classe `connectivity` ce l'hanno il router, la stampante e ogni telefono
  /// — e il registro si'. Arriva solo quando lo si chiede per nome.
  final String piattaforma;

  /// `total`, `total_increasing`, `measurement`. Distingue un contatore che
  /// sale per sempre da una lettura che va su e giu': senza, il «totale» di un
  /// elettrodomestico e' l'energia di oggi una volta su due.
  final String classeDiStato;

  /// `config` o `diagnostic`: entita' di servizio, che di solito non si
  /// mettono in una tessera.
  final String categoria;

  final bool spenta;
}

/// Quello che il ponte risponde.
class IlCatalogo {
  const IlCatalogo({
    required this.integrazioni,
    required this.dispositivi,
    required this.entita,
  });

  const IlCatalogo.vuoto()
    : integrazioni = const [],
      dispositivi = const [],
      entita = const {};

  final List<Integrazione> integrazioni;
  final List<Dispositivo> dispositivi;

  /// Le entita', per dispositivo. Arrivano solo per i dispositivi che si
  /// chiedono: una casa grande ne ha migliaia, e il menu ne vuole vedere venti
  /// alla volta.
  final Map<String, List<EntitaDelDispositivo>> entita;

  List<Dispositivo> dellIntegrazione(String dominio) => [
    for (final uno in dispositivi)
      if (uno.integrazione == dominio) uno,
  ];
}

/// Chiede il catalogo al ponte.
///
/// [dispositivi] chiede anche le entita' di quei dispositivi. Senza, arrivano
/// solo le integrazioni e i dispositivi — che e' quello che serve ai primi due
/// passi del menu.
Future<IlCatalogo> chiediIlCatalogo(
  Filo filo, {
  List<String> dispositivi = const [],
  List<String> entita = const [],
}) async {
  final detto = await filo.chiedi({
    'type': comandoDelCatalogo,
    if (dispositivi.isNotEmpty) 'device_ids': dispositivi,
    /* Per nome: di quale integrazione e' questa entita'. E' la domanda che
     * fa la scheda delle macchine, e il registro e' l'unico che lo sa. */
    if (entita.isNotEmpty) 'entity_ids': entita,
  }, entro: attesaDelCatalogo);
  return leggiIlCatalogo(detto['result']);
}

/// Separato da chi lo chiede, cosi' le prove non hanno bisogno di un filo.
IlCatalogo leggiIlCatalogo(dynamic risultato) {
  if (risultato is! Map) return const IlCatalogo.vuoto();
  /* Un campo che non e' un elenco vale come un elenco vuoto, non come un
   * errore: il catalogo arriva da fuori, e una risposta storta deve lasciare
   * il menu vuoto, non farlo cadere. Il cast diretto invece sollevava. */
  List<Map<String, dynamic>> elenco(String dove) {
    final letto = risultato[dove];
    if (letto is! List) return const [];
    return [
      for (final uno in letto)
        if (uno is Map) Map<String, dynamic>.from(uno),
    ];
  }

  final entita = <String, List<EntitaDelDispositivo>>{};
  for (final una in elenco('entities')) {
    /* Chieste per nome possono non avere un dispositivo: stanno sotto la
     * chiave vuota, che e' comunque una risposta. */
    final quale = '${una['device_id'] ?? ''}';
    (entita[quale] ??= []).add(
      EntitaDelDispositivo(
        id: '${una['entity_id'] ?? ''}',
        nome: '${una['name'] ?? una['entity_id'] ?? ''}',
        classe: '${una['device_class'] ?? ''}',
        unita: '${una['unit'] ?? ''}',
        categoria: '${una['category'] ?? ''}',
        spenta: una['disabled'] == true || una['hidden'] == true,
        chiaveDiTraduzione: '${una['translation_key'] ?? ''}',
        classeDiStato: '${una['state_class'] ?? ''}',
        piattaforma: '${una['platform'] ?? ''}',
      ),
    );
  }

  return IlCatalogo(
    integrazioni: [
      for (final una in elenco('integrations'))
        Integrazione(
          dominio: '${una['domain'] ?? ''}',
          nome: '${una['name'] ?? una['domain'] ?? ''}',
          quantiDispositivi: (una['devices'] as num?)?.toInt() ?? 0,
          /* `custom` puo' essere `null` quando il manifesto non c'e': allora
           * non si sa, e non si dice. */
          diQualcunAltro: una['custom'] == true,
        ),
    ],
    dispositivi: [
      for (final uno in elenco('devices'))
        Dispositivo(
          id: '${uno['id'] ?? ''}',
          nome: '${uno['name'] ?? ''}',
          marca: '${uno['manufacturer'] ?? ''}',
          modello: '${uno['model'] ?? ''}',
          integrazione: '${uno['integration'] ?? ''}',
          stanza: '${uno['area'] ?? ''}',
          quanteEntita: (uno['entities'] as num?)?.toInt() ?? 0,
          spento: uno['disabled'] == true,
        ),
    ],
    entita: entita,
  );
}
